### ISSUE TITLE
An infinite loop of implicit convertions

### ISSUE THREAD

**Issue by KernelErr0r** [dotnet/roslyn#43264]
This smart hack to remove generics makes an infinite loop of implicit conversions. I think it shouldn't call the same operator twice in a row. Is it some kind of a bug, unpredicted behavior or does it have to work that way?
```cs
public class Program
{
	public static void Main()
	{
		Module module = new Module<string>();
	}
}

public class Module : Module<object>
{

}

public class Module<T>
{
	public static implicit operator Module(Module<T> module)
	{
		return module;
	}
}
```

**Comment by sharwell** [comment]
@CyrusNajmabadi should we create an analyzer for this?

**Comment by CyrusNajmabadi** [comment]
@sharwell  I don't think Roslyn should.  But i think Roslyn-analyzers should.  @mavasani @jinujoseph  would it be in scope for that projecT?  If so, let's move over.  If not, we could pick it up :)

**Comment by CyrusNajmabadi** [comment]
>  Is it some kind of a bug, unpredicted behavior or does it have to work that way?

It's technically by-design as far as the language is concerned.  But it's certainly confusing, unexpected, and has tripped up many people.  So an analyzer to detect infinite calls makes a lot of sense.

**Comment by gafter** [comment]
The compiler should warn (in a warning wave) when an implicit conversion declaration uses itself.

**Comment by YairHalberstadt** [comment]
> The compiler should warn (in a warning wave) when an implicit conversion declaration uses itself.

Why is this different to any other recursive call? It might well be by design:

```csharp
using System;
public abstract class Tree<T>
{
    private Tree(){}
    
    public class Leaf : Tree<T>
    {
        public T Value { get; set; }
    }
    
    public class Branch : Tree<T>
    {
        public Tree<T> Left { get; set; }
        public Tree<T> Right { get; set; }
    }
    
    public static implicit operator Tree(Tree<T> tree)
    {
        if (tree is Leaf leaf)
            return new Tree.Leaf{Value = leaf.Value };
        if (tree is Branch branch)
            return new Tree.Branch { Left = branch.Left, Right = branch.Right };
        throw new InvalidOperationException();
    }
}

public abstract class Tree
{
    private Tree(){}
    
    public class Leaf : Tree
    {
        public object Value { get; set; }
    }
    
    public class Branch : Tree
    {
        public Tree Left { get; set; }
        public Tree Right { get; set; }
    }
}
```

**Comment by CyrusNajmabadi** [comment]
> Why is this different to any other recursive call?

Primarily because it is implicit.  i.e. when you have a recursive call, you're doing:

```c#
void Foo()
{
  return Foo(); //<-- i can at least see this
}
```

When you have:

```c#
	public static implicit operator Module(Module<T> module)
	{
		return module;
	}
```

You really have:

```
	public static implicit operator Module(Module<T> module)
	{
                // effectively:
		return Module<T>.op_implicit<Module>(module);
	}
```

But it's not at all clear that's what's going on.  Note: i would be fine with an analyzer that detected obviously wrong self-recursion.  i.e. the direct return/call of the exact method with the exact same args.  In that case, this would be the same as normal method recursion in terms of getting warnings.

**Comment by YairHalberstadt** [comment]
> Note: i would be fine with an analyzer that detected obviously wrong self-recursion. i.e. the direct return/call of the exact method with the exact same args.

I think that would make more sense. Getting a warning for the above tree example would be annoying.
